packages/abi/src/parser/specifications/v1/resolvable-type.ts

```diff
@@ -0,0 +1,269 @@
+import { swayTypeMatchers } from '../../../matchers/sway-type-matchers';
+import type { AbiTypeMetadata } from '../../abi';
+
+import { ResolvedType } from './resolved-type';
+import type {
+  AbiComponentV1,
+  AbiConcreteTypeV1,
+  AbiMetadataTypeV1,
+  AbiSpecificationV1,
+  AbiTypeArgumentV1,
+} from './specification';
+
+interface ResolvableComponent {
+  name: string;
+  type: ResolvableType | ResolvedType;
+}
+
+export class ResolvableType {
+  private metadataType: AbiMetadataTypeV1;
+  type: string;
+  components: ResolvableComponent[] | undefined;
+
+  toAbiType(): AbiTypeMetadata {
+    return {
+      metadataTypeId: this.metadataTypeId,
+      components: this.components?.map((c) => ({ name: c.name, type: c.type.toAbiType() })),
+      swayType: this.type,
+      typeParameters: this.typeParamsArgsMap?.map(([, t]) => (t as ResolvableType).toAbiType()),
+    };
+  }
+
+  constructor(
+    private abi: AbiSpecificationV1,
+    public metadataTypeId: number,
+    public typeParamsArgsMap: Array<[number, ResolvedType | ResolvableType]> | undefined
+  ) {
+    const metadataType = abi.metadataTypes.find(
+      (tm) => tm.metadataTypeId === metadataTypeId
+    ) as AbiMetadataTypeV1;
+
+    this.metadataType = metadataType;
+    this.type = metadataType.type;
+
+    this.typeParamsArgsMap ??= ResolvableType.mapTypeParametersAndArgs(
+      abi,
+      metadataType,
+      undefined
+    );
+
+    let components = metadataType.components;
+
+    /**
+     * Vectors consist of multiple components,
+     * but we only care about the `buf`'s first type argument
+     * which defines the type of the vector data.
+     * Everything else is being ignored,
+     * as it's then easier to reason about the vector
+     * (you just treat is as a regular struct).
+     */
+    if (swayTypeMatchers.vector(metadataType.type)) {
+      components = components?.map((c) => {
+        if (c.name === 'buf') {
+          return c.typeArguments?.[0];
+        }
+        return c;
+      }) as AbiComponentV1[];
+    }
+
+    this.components = components?.map((c) => ResolvableType.handleComponent(this, abi, c));
+  }
+
+  private static mapTypeParametersAndArgs(
+    abi: AbiSpecificationV1,
+    metadataType: AbiMetadataTypeV1,
+    args: (ResolvableType | ResolvedType)[] | undefined
+  ): Array<[number, ResolvedType | ResolvableType]> | undefined {
+    if (!args) {
+      return metadataType.typeParameters?.map((typeParameter) => [
+        typeParameter,
+        new ResolvableType(abi, typeParameter, undefined),
+      ]);
+    }
+
+    return metadataType.typeParameters?.map((typeParameter, idx) => [typeParameter, args[idx]]);
+  }
+
+  private static handleComponent(
```

Comment:
I believe we can rename this method to `createResolvableComponent` as it reflects its purpose clearly